### PR TITLE
feat(pdf): inject vimium-c (#1462)

### PR DESCRIPTION
**Pull request by gdh1995**
This is a basic version to inject Vimium C by importing `vimium-c-injector.js`, which is modified from the one in EdgeTranslate.

The differences are:
* use promise-style APIs
* add detection of global `browser` instance and use it
* add the ID of Vimium C (`vimium-c@gdh1995.cn`) on Firefox Add-ons

### CODE REVIEW

**Comment by gdh1995** [comment]
There's a known issue in the JS file: it always reads (and updates) `storage.sync`. It's because I don't know how to get an expected storage area from user configuration. So you may modify this PR directly if only needed.

File: assets/vimium-c-injector.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        if (injector === "nul" || injector === "/dev/null") {
+            return;
+        }
+        if (!injector) {
+            injector = /\sEdg\//.test(navigator.appVersion)
+                ? "aibcglbfblnogfjhbcmmpobjhnomhcdo"
+                : typeof InstallTrigger !== "undefined"
+                ? "vimium-c@gdh1995.cn"
+                : "hfjbmagddngcpeloejdejnfgbamkjaeg";
+            browser_.storage.sync.set({ vimiumExtensionInjector: injector });
```
Comment (crimx) [inline assets/vimium-c-injector.js]: 为什么这里需要存起来？

File: assets/vimium-c-injector.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                box.scrollTo(position[0], position[1]);
+                channelElement.textContent = "";
+                event.preventDefault();
+            }
+        } else {
+            // 对应命令 Marks.activateCreate
+            channelElement.textContent = [box.scrollLeft, box.scrollTop];
+        }
+    };
+    const browser_ = typeof browser === "object" && browser && browser.runtime ? browser : chrome
```
Comment (crimx) [inline assets/vimium-c-injector.js]: 直接使用 browser 即可，已经 polyfill 过

**Review by crimx — CHANGES_REQUESTED** [review]

**Comment by crimx** [comment]
Thanks for the PR! 

> add detection of global browser instance and use it

`browser` is polyfilled in Saladict. You can use it directly.

> add the ID of Vimium C (vimium-c@gdh1995.cn) on Firefox Add-ons

Do we also need to add CSP to the Firefox manifest.json?

> There's a known issue in the JS file: it always reads (and updates) storage.sync

Why is this needed?

**Comment by gdh1995** [comment]
第1个是有两个目的：
1. 为了方便嗯想改id的人，比如说他从别的地方弄了个vimium c，想知道怎么样用自己的id而不是商店版本的，就查一下storage里的数据就行了。
2. 这个脚本其实是从原来某个脚本改出来的，之前在edge上会尝试edge和chrome两个商店里的id，嗯，然后尝试成功之后就存下来，下次加载不用再试了。嗯，当然现在这个版本里面没有这个功能了


第2个的话，嗯确实应该直接用browser



---Original---
From: ***@***.***&gt;
Date: Thu, Sep 2, 2021 21:11 PM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [crimx/ext-saladict] feat(pdf): inject vimium-c (#1462) (#1463)




 
@crimx requested changes on this pull request.
 
 
In assets/vimium-c-injector.js:
 &gt; +        } +    }; +    const browser_ = typeof browser === "object" &amp;&amp; browser &amp;&amp; browser.runtime ? browser : chrome + +    browser_.storage.sync.get("vimiumExtensionInjector").then((result) =&gt; { +        let injector = result.vimiumExtensionInjector; +        if (injector === "nul" || injector === "/dev/null") { +            return; +        } +        if (!injector) { +            injector = /\sEdg\//.test(navigator.appVersion) +                ? "aibcglbfblnogfjhbcmmpobjhnomhcdo" +                : typeof InstallTrigger !== "undefined" +                ? ***@***.***" +                : "hfjbmagddngcpeloejdejnfgbamkjaeg"; +            browser_.storage.sync.set({ vimiumExtensionInjector: injector });  
为什么这里需要存起来？
 
 
In assets/vimium-c-injector.js:
 &gt; +        const str = channelElement.textContent; +        if (str) { +            // 对应命令 Marks.activate +            const mark = str.split(","); +            const position = [~~mark[0], ~~mark[1]]; +            if (position[0] &gt; 0 || position[1] &gt; 0) { +                box.scrollTo(position[0], position[1]); +                channelElement.textContent = ""; +                event.preventDefault(); +            } +        } else { +            // 对应命令 Marks.activateCreate +            channelElement.textContent = [box.scrollLeft, box.scrollTop]; +        } +    }; +    const browser_ = typeof browser === "object" &amp;&amp; browser &amp;&amp; browser.runtime ? browser : chrome  
直接使用 browser 即可，已经 polyfill 过
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by gdh1995** [comment]
Firefox上不需要csp。


嗯，我的意思是说他一直会读sync，而不是local，这一点不确定要不要改、要怎么改。至于往存储里写的地方，嗯，也可以直接删的。



---Original---
From: ***@***.***&gt;
Date: Thu, Sep 2, 2021 21:30 PM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [crimx/ext-saladict] feat(pdf): inject vimium-c (#1462) (#1463)




 
Thanks for the PR!
  
add detection of global browser instance and use it
  
browser is polyfilled in Saladict. You can use it directly.
  
add the ID of Vimium C ***@***.***) on Firefox Add-ons
  
Do we also need to add CSP to the Firefox manifest.json?
  
There's a known issue in the JS file: it always reads (and updates) storage.sync
  
Why is this needed?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by crimx** [comment]
> 为了方便嗯想改id的人，比如说他从别的地方弄了个vimium c，想知道怎么样用自己的id而不是商店版本的，就查一下storage里的数据就行了。

但假如ta改了 id，CSP 怎么处理？

**Comment by gdh1995** [comment]
没法处理，要么是firefox上不需要csp，要么是也要改saladict



---Original---
From: ***@***.***&gt;
Date: Thu, Sep 2, 2021 21:41 PM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [crimx/ext-saladict] feat(pdf): inject vimium-c (#1462) (#1463)




  
为了方便嗯想改id的人，比如说他从别的地方弄了个vimium c，想知道怎么样用自己的id而不是商店版本的，就查一下storage里的数据就行了。
  
但假如ta改了 id，CSP 怎么处理？
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by gdh1995** [comment]
嗯，那这一行还是删掉吧



---Original---
From: ***@***.***&gt;
Date: Thu, Sep 2, 2021 21:41 PM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [crimx/ext-saladict] feat(pdf): inject vimium-c (#1462) (#1463)




  
为了方便嗯想改id的人，比如说他从别的地方弄了个vimium c，想知道怎么样用自己的id而不是商店版本的，就查一下storage里的数据就行了。
  
但假如ta改了 id，CSP 怎么处理？
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: assets/vimium-c-injector.js (hunk 1 above)
Comment (gdh1995) [reply assets/vimium-c-injector.js]: 嗯新提交里改了，browser_也换成browser了。另外加了在Edge上自动尝试连接Chrome商店版Vimium C的功能。

File: assets/vimium-c-injector.js [hunk 3]
```diff
@@ -72,44 +72,53 @@
...
         }
     };
-    const browser_ = typeof browser === "object" && browser && browser.runtime ? browser : chrome
 
-    browser_.storage.sync.get("vimiumExtensionInjector").then((result) => {
+    const IDOnFirefox = "vimium-c@gdh1995.cn";
+    const IDOnEdge = "aibcglbfblnogfjhbcmmpobjhnomhcdo";
+    const IDOnChrome = "hfjbmagddngcpeloejdejnfgbamkjaeg";
+
+    browser.storage.sync.get("vimiumExtensionInjector").then((result) => {
```
Comment (crimx) [inline assets/vimium-c-injector.js:80]: 这个 vimiumExtensionInjector 是不是也可以去掉？

File: src/manifest/chrome.manifest.json [hunk 4]
```diff
@@ -8,6 +8,7 @@
   "optional_permissions": [
     "background"
   ],
+  "content_security_policy": "script-src 'self' 'unsafe-eval' chrome-extension://hfjbmagddngcpeloejdejnfgbamkjaeg/ chrome-extension://aibcglbfblnogfjhbcmmpobjhnomhcdo/; object-src 'self'",
```
Comment (crimx) [inline src/manifest/chrome.manifest.json]: 你好，发现需要的权限还是有点大，为什么 vimium-c 需要 unsafe-eval ？沙拉查词原来需要是因为 PDF.js 的 polyfill，现在估计也要去掉。

**Review by crimx — CHANGES_REQUESTED** [review]

File: assets/vimium-c-injector.js (hunk 3 above)
Comment (gdh1995) [reply assets/vimium-c-injector.js:80]: 去掉的话用户就没法关掉这个功能了。只是在 Vimium C 的白名单里去掉saladict的ID的话，pdf页面控制台里会打日志说connection refused。或者把这个日志也去掉？

File: src/manifest/chrome.manifest.json (hunk 4 above)
Comment (gdh1995) [reply src/manifest/chrome.manifest.json]: 确定能去掉吗？目前版本的 https://github.com/crimx/ext-saladict/blob/452bf537a4aee7d1e3d79b7960cd5183cdff88af/src/manifest/common.manifest.js#L98 里还有这个，我照搬了；原版 pdf.js 里也有（ https://github.com/mozilla/pdf.js/blob/d95f680d66dd46e2995a7784109a1b3af2b84bdd/extensions/chromium/manifest.json#L31 ）。

Comment (crimx) [reply src/manifest/chrome.manifest.json]: @gdh1995 原来提交 Firefox 的时候被审核要求去过，排查了一下发现是 PDF.js 的一些 ES5 polyfill 需要这个跑，移除了 polyfill 发现也没问题。现在 PDF.js 打包可以直接输出 es6 了，所以应该就没有这个问题了。Chrome 这边我还没删掉。

Comment (gdh1995) [reply src/manifest/chrome.manifest.json]: 加了一个提交，把 unsafe-eval去掉了

Comment (gdh1995) [reply src/manifest/chrome.manifest.json]: 呃，看起来 eaea459ae500cf84cea3f65e59c573d6816d222d (2020-04-28) 去掉了某些pdf.js的代码，但是 f7d9f2c3521e57be62bd3296353ff2a75798fa69 (2021-05-27) 回滚了这些改动。

Comment (gdh1995) [reply src/manifest/chrome.manifest.json]: 哦没事，我看了看pdf.js的内容，确实已经不需要了

**Comment by crimx** [comment]
🐂🍺！

**Pull request merged**